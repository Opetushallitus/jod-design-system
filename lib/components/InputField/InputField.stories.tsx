import { useState } from '@storybook/preview-api';
import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';
import { InputField } from './InputField';

const meta = {
  title: 'Form/InputField',
  component: InputField,
  tags: ['autodocs'],
} satisfies Meta<typeof InputField>;

export default meta;

type Story = StoryObj<typeof meta>;

const render = (args: Story['args']) => {
  const { value, onChange, ...rest } = args;
  const [textValue, setTextValue] = useState(value);
  return (
    <InputField
      value={textValue}
      onChange={(event) => {
        setTextValue(event.target.value);
        onChange(event);
      }}
      {...rest}
    />
  );
};
const url = 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-7019&t=RZCfmzkZuXUFBRkG-4';
const label = 'Label text';
const loremIpsum = 'Lorem ipsum dolor sit amet';

export const Default: Story = {
  render,
  decorators: [
    (Story) => (
      <div className="ds:max-w-[415px]">
        <Story />
      </div>
    ),
  ],
  parameters: {
    design: {
      type: 'figma',
      url,
    },
    docs: {
      description: {
        story: 'This is a input field component for displaying and editing a value with a label text.',
      },
    },
  },
  args: {
    value: loremIpsum,
    onChange: fn(),
    label,
  },
};

export const HelpText: Story = {
  render,
  decorators: [
    (Story) => (
      <div className="ds:max-w-[415px]">
        <Story />
      </div>
    ),
  ],
  parameters: {
    design: {
      type: 'figma',
      url,
    },
    docs: {
      description: {
        story: 'This is a input field component for displaying and editing a value with a label and help text.',
      },
    },
  },
  args: {
    value: loremIpsum,
    onChange: fn(),
    label,
    help: 'Help text',
  },
};

export const Placeholder: Story = {
  render,
  decorators: [
    (Story) => (
      <div className="ds:max-w-[415px]">
        <Story />
      </div>
    ),
  ],
  parameters: {
    design: {
      type: 'figma',
      url,
    },
    docs: {
      description: {
        story: 'This is a input field component for displaying and editing a value with placeholder text.',
      },
    },
  },
  args: {
    value: '',
    onChange: fn(),
    label,
    placeholder: loremIpsum,
  },
};
