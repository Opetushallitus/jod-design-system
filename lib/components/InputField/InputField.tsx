import React from 'react';

export interface InputFieldProps {
  /** The name of the input field */
  name?: string;
  /** The value of the input field */
  value?: string;
  /** The function to call when the input field loses focus */
  onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
  /** The function to call when the value of the input field changes */
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  /** The placeholder text to display in the input field */
  placeholder?: string;
  /** The label text to display above the input field */
  label: string;
  /** The help text to display below the input field */
  help?: string;
}

/** Input fields are text boxes that allow users to input custom text entries with a keyboard. Various options can be shown with the field to communicate the input requirements. */
export const InputField = React.forwardRef<HTMLInputElement, InputFieldProps>(function InputField(
  { name, value, onBlur, onChange, placeholder, label, help }: InputFieldProps,
  ref,
) {
  const inputId = React.useId();
  const helpId = React.useId();
  return (
    <>
      <label htmlFor={inputId} className="ds-mb-4 ds-inline-block ds-align-top ds-text-form-label ds-text-black">
        {label}
      </label>
      <input
        ref={ref}
        id={inputId}
        name={name}
        type="text"
        value={value}
        onBlur={onBlur}
        onChange={onChange}
        placeholder={placeholder ? `(${placeholder})` : undefined}
        autoComplete="off"
        aria-describedby={help ? helpId : undefined}
        className="ds-block ds-w-full ds-rounded ds-border ds-border-border-gray ds-bg-white ds-p-[11px] ds-text-black ds-outline-none placeholder:ds-text-secondary-gray"
      />
      {help && (
        <div id={helpId} className="ds-mt-2 ds-block ds-text-help ds-text-secondary-gray">
          {help}
        </div>
      )}
    </>
  );
});
