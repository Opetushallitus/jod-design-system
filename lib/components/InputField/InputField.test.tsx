import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import { InputField } from './InputField';

describe('InputField', () => {
  const label = 'Username';

  it('renders label correctly', () => {
    const value = 'john';
    const onChange = vi.fn();
    const { container } = render(<InputField value={value} onChange={onChange} label={label} />);
    const labelElement = screen.getByText(label);
    expect(labelElement).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders input correctly', () => {
    const placeholder = 'Enter your username';
    const value = 'john';
    const onChange = vi.fn();
    const { container } = render(
      <InputField value={value} onChange={onChange} label={label} placeholder={placeholder} />,
    );
    const inputElement = screen.getByPlaceholderText(`(${placeholder})`);
    expect(inputElement).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders help text correctly', () => {
    const helpText = 'Please enter your username';
    const value = 'john';
    const onChange = vi.fn();
    const { container } = render(<InputField value={value} onChange={onChange} label={label} help={helpText} />);
    const helpElement = screen.getByText(helpText);
    expect(helpElement).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('updates input value correctly', () => {
    const onChange = vi.fn();
    const value = 'test value';
    const { container } = render(<InputField value={value} onChange={onChange} label={label} />);
    const inputElement = screen.getByRole('textbox');
    fireEvent.change(inputElement, { target: { value: 'new value' } });
    expect(onChange).toHaveBeenCalled();
    expect(container.firstChild).toMatchSnapshot();
  });
});
