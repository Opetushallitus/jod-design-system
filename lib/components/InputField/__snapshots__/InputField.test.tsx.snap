// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputField > renders help text correctly 1`] = `
<div>
  <label
    class="ds:mb-4 ds:inline-block ds:align-top ds:text-form-label ds:font-arial ds:text-primary-gray"
    for="«r4»"
  >
    Username
  </label>
  <input
    aria-describedby="«r5»"
    aria-required="false"
    autocomplete="off"
    class="ds:block ds:w-full ds:rounded ds:border ds:border-border-gray ds:bg-white ds:p-5 ds:text-primary-gray ds:focus:outline-2 ds:focus:outline-accent ds:placeholder:text-inactive-gray ds:font-arial ds:text-body-md"
    id="«r4»"
    type="text"
    value="john"
  />
  <div
    class="ds:mt-2 ds:block ds:text-help ds:text-secondary-gray ds:font-arial"
    id="«r5»"
  >
    Please enter your username
  </div>
</div>
`;

exports[`InputField > renders input correctly 1`] = `
<div>
  <label
    class="ds:mb-4 ds:inline-block ds:align-top ds:text-form-label ds:font-arial ds:text-primary-gray"
    for="«r2»"
  >
    Username
  </label>
  <input
    aria-required="false"
    autocomplete="off"
    class="ds:block ds:w-full ds:rounded ds:border ds:border-border-gray ds:bg-white ds:p-5 ds:text-primary-gray ds:focus:outline-2 ds:focus:outline-accent ds:placeholder:text-inactive-gray ds:font-arial ds:text-body-md"
    id="«r2»"
    placeholder="Enter your username"
    type="text"
    value="john"
  />
</div>
`;

exports[`InputField > renders label correctly 1`] = `
<div>
  <label
    class="ds:mb-4 ds:inline-block ds:align-top ds:text-form-label ds:font-arial ds:text-primary-gray"
    for="«r0»"
  >
    Username
  </label>
  <input
    aria-required="false"
    autocomplete="off"
    class="ds:block ds:w-full ds:rounded ds:border ds:border-border-gray ds:bg-white ds:p-5 ds:text-primary-gray ds:focus:outline-2 ds:focus:outline-accent ds:placeholder:text-inactive-gray ds:font-arial ds:text-body-md"
    id="«r0»"
    type="text"
    value="john"
  />
</div>
`;

exports[`InputField > updates input value correctly 1`] = `
<div>
  <label
    class="ds:mb-4 ds:inline-block ds:align-top ds:text-form-label ds:font-arial ds:text-primary-gray"
    for="«r6»"
  >
    Username
  </label>
  <input
    aria-required="false"
    autocomplete="off"
    class="ds:block ds:w-full ds:rounded ds:border ds:border-border-gray ds:bg-white ds:p-5 ds:text-primary-gray ds:focus:outline-2 ds:focus:outline-accent ds:placeholder:text-inactive-gray ds:font-arial ds:text-body-md"
    id="«r6»"
    type="text"
    value="test value"
  />
</div>
`;
